designs/build_debug_artifacts.md

```diff
@@ -0,0 +1,120 @@
+SAM Build producing debuggable artifacts
+========================================
+
+What is the problem?
+--------------------
+
+`sam build` produces artifacts that are built for production use. In some langauges (usually interpreted), 
+these production artifacts are also debuggable locally. In the case of compiled languages, you usually need to compile
+the binary or artifact in a specific manner for them to be debuggable.  
+
+What will be changed?
+---------------------
+
+We will introduce a way in `sam build` to produce these debuggable artifacts for those compiled languages.
+
+Success criteria for the change
+-------------------------------
+
+1. Artifacts generated will be debuggable for DotNetCore2.1 Runtimes
+
+Out-of-Scope
+------------
+
+1. Other languages `sam build` supports will not be changed
+
+User Experience Walkthrough
+---------------------------
+
+Implementation
+==============
+
+CLI Changes
+-----------
+
+*Explain the changes to command line interface, including adding new
+commands, modifying arguments etc*
+
+### Breaking Change
+
+Changes are additive and will not break any existing functionality.
+
+Design
+------
+
+Options considered
+------------------
+We have a couple options to consider:
+
+1. A command line option (`--debug-artifact`)
+    * Pros
+        * Customer can control what kind of artifacts are produced.
+        * There is no guessing needed by the CLI
+    * Cons
+        * Yet another option for customers to learn or know about
+        * Makes the customer need to think about the artifacts they need to produce (though this can be hidden
+        behind the AWS Toolkit in IDEs)
+        * Customers running in the command line need to remember what artifacts to produce or what they previously produced
+2. An Environment Variable we read (`SAM_BUILD_DEBUG_ARTIFACT`)
+    * Pros
+        * Could force customers to use AWS Toolkit by default
+    * Cons
+        * Building debug artifacts becomes a hidden feature (silent contract)
+        * Environment Variables tend to be more set and forget.
+        * Need to conditionally add Env Var instead of a more convenient flag
+3. `sam build` produces debug artifacts by default and `sam package` will build non debug artifacts by default
+    * Pros
+        * Customers should never have to think about 'when to produce debug artifacts' or forget to add a flag
```

Comment:
They are completely different usecases right? so I would think would need to know about the modes required to produce 2 different sets of artifacts.

Reply:
Maybe? The framing I was thinking is, if one command does the building, packaging, and deploying then the action is defined for us. 

So let's just say we have two commands `invoke` (which only builds and invokes) and `deploy` (which builds, packages, and deploys). One could argue that `invoke` would always build debuggable artifacts and hide away the "do I need do pass different options to the command?". `Deploy` on the other hand, could build for production (leaving out debug headers, etc). In this model, customers never have to think about how they need to build. They would only have to care about what action do I need to accomplish. This can simplify customers interactions, of course we still need some way of controlling things (for say AWS Toolkit and any other tools) and there are other nuances I did not take into consideration. Strictly from a CLI point of view though, we may not need an option at all and that is what this was attempting to capture.